lc0/src/neural/network_opencl.cc

```diff
@@ -0,0 +1,447 @@
+/*
+ This file is part of Leela Chess Zero.
+ Copyright (C) 2018 The LCZero Authors
+ 
+ Leela Chess is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+ 
+ Leela Chess is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "neural/factory.h"
+#include "neural/network.h"
+
+#include <condition_variable>
+#include <cassert>
+#include <iostream>
+#include <cmath>
+#include <algorithm>
+#include <thread>
+#include "utils/exception.h"
+
+#include "CL/OpenCLUtils.h"
+
+#include "CL/OpenCLParams.h"
+#include "CL/OpenCL.h"
+
+#ifdef __APPLE__
+#include <Accelerate/Accelerate.h>
+#endif
+#ifdef USE_MKL
+#include <mkl.h>
+#endif
+#ifdef USE_OPENBLAS
+#include <cblas.h>
+#endif
+
+
+namespace lczero {
+  
+  namespace {
+    
+    
+    static constexpr int NUM_VALUE_INPUT_PLANES = 32;
+    static constexpr int NUM_POLICY_INPUT_PLANES = 32;
+    
+    static constexpr int NUM_OUTPUT_POLICY = 1858;
+    static constexpr int NUM_VALUE_CHANNELS = 128;
+    
+    
+    class OpenCLNetwork;
+    
+    // Copy the vectors we need after weights is deallocated
+    struct OpenCLWeights {
```

Comment:
I don't think this name is very good, since it's just the weights that *aren't* used by OpenCL

Reply:
Yes, I thought of it. You are right, we could use a non-misleading name and more self-explaining name. I was out of ideas and I thought the comments line would be enough.

I you have a suggestion of good name, tell me.